### DISCUSSION TITLE
Authelia oidc  with centreon

### DISCUSSION THREAD

**Discussion by g4fa** [authelia/authelia#6466] [Q&A] [answered]
Hi 

I try to configure oidc  Authelia with centreon. Is some one successful ton configure Authelia with centreon ? 

In  /Var/log/centreon/login.log , I have this error : "Openid] [Error] Unable to get Token Access Information: Core\Security\Authentication\Domain\Exception\SSOAuthenticationException, message: Request for connection token to external provider has failed"

        - id: centreon
          description: centreon
          secret: xxxxxxxxxxxx  # The digest of 'insecure_secret'.
          public: false
          authorization_policy: two_factor
          redirect_uris:
            - https://xxxxxxxxxx/centreon/authentication/providers/configurations/openid
          scopes:
            - openid
            - profile
          userinfo_signing_algorithm: none

Thanks for your help 
Fabien

**Comment by james-d-elliott** [comment]
Sounds like a issue with networking. Why are you not letting us know what they said about the error message?

**Reply by g4fa** [reply]
![image](https://github.com/authelia/authelia/assets/20145195/12bce8f1-c490-4121-9b74-54a65ac05e46)
this my configuration in Centreon

And the log 
time="2023-12-27T14:05:32+01:00" level=debug msg="Authorization Request with id '9a43522c-f21e-4757-8f17-548477735fa5' on client with id 'centreon' is being processed" method=GET path=/api/oidc/authorization remote_ip=xxxx.xxx.xxx.xxx
time="2023-12-27T14:05:32+01:00" level=debug msg="Authorization Request with id '9a43522c-f21e-4757-8f17-548477735fa5' on client with id 'centreon' using consent mode 'explicit' proceeding to generate a new consent session" method=GET path=/api/oidc/authorization remote_ip=xxxx.xxx.xxx.xxx
time="2023-12-27T14:05:32+01:00" level=debug msg="Authorization Request with id '9a43522c-f21e-4757-8f17-548477735fa5' on client with id 'centreon' using consent mode 'explicit' authentication level 'two_factor' is sufficient for client level 'two_factor'" method=GET path=/api/oidc/authorization remote_ip=xxxx.xxx.xxx.xxx
time="2023-12-27T14:05:32+01:00" level=debug msg="Authorization Request with id '9a43522c-f21e-4757-8f17-548477735fa5' on client with id 'centreon' using consent mode 'explicit' is being redirected to 'https://authelia.xxx-xxx.fr/consent?id=039b2775-ac7d-4053-9e5e-e5ddf7086598'" method=GET path=/api/oidc/authorization remote_ip=xxxx.xxx.xxx.xxx
time="2023-12-27T14:05:32+01:00" level=trace msg="Replied (status=302)" method=GET path=/api/oidc/authorization remote_ip=xxxx.xxx.xxx.xxx
time="2023-12-27T14:05:32+01:00" level=trace msg="Request hit" method=GET path=/consent remote_ip=xxxx.xxx.xxx.xxx
time="2023-12-27T14:05:32+01:00" level=trace msg="Replied (status=200)" method=GET path=/consent remote_ip=xxxx.xxx.xxx.xxx
time="2023-12-27T14:05:32+01:00" level=trace msg="Request hit" method=GET path=/manifest.json remote_ip=xxxx.xxx.xxx.xxx
time="2023-12-27T14:05:32+01:00" level=trace msg="Replied (status=200)" method=GET path=/manifest.json remote_ip=xxxx.xxx.xxx.xxx
time="2023-12-27T14:05:32+01:00" level=trace msg="Request hit" method=GET path=/locales/fr-FR/portal.json remote_ip=xxxx.xxx.xxx.xxx
time="2023-12-27T14:05:32+01:00" level=trace msg="Replied (status=200)" method=GET path=/locales/fr-FR/portal.json remote_ip=xxxx.xxx.xxx.xxx
time="2023-12-27T14:05:32+01:00" level=trace msg="Request hit" method=GET path=/locales/fr/portal.json remote_ip=xxxx.xxx.xxx.xxx
time="2023-12-27T14:05:32+01:00" level=trace msg="Replied (status=200)" method=GET path=/locales/fr/portal.json remote_ip=xxxx.xxx.xxx.xxx
time="2023-12-27T14:05:32+01:00" level=trace msg="Request hit" method=GET path=/locales/en/portal.json remote_ip=xxxx.xxx.xxx.xxx
time="2023-12-27T14:05:32+01:00" level=trace msg="Replied (status=200)" method=GET path=/locales/en/portal.json remote_ip=xxxx.xxx.xxx.xxx
time="2023-12-27T14:05:32+01:00" level=trace msg="Request hit" method=GET path=/api/user/info remote_ip=xxxx.xxx.xxx.xxx
time="2023-12-27T14:05:32+01:00" level=trace msg="Replied (status=200)" method=GET path=/api/user/info remote_ip=xxxx.xxx.xxx.xxx
time="2023-12-27T14:05:32+01:00" level=trace msg="Request hit" method=GET path=/api/oidc/consent remote_ip=xxxx.xxx.xxx.xxx
time="2023-12-27T14:05:32+01:00" level=trace msg="Replied (status=200)" method=GET path=/api/oidc/consent remote_ip=xxxx.xxx.xxx.xxx
time="2023-12-27T14:05:33+01:00" level=trace msg="Request hit" method=POST path=/api/oidc/consent remote_ip=xxxx.xxx.xxx.xxx
time="2023-12-27T14:05:33+01:00" level=trace msg="Replied (status=200)" method=POST path=/api/oidc/consent remote_ip=xxxx.xxx.xxx.xxx
time="2023-12-27T14:05:33+01:00" level=trace msg="Request hit" method=GET path=/api/oidc/authorization remote_ip=xxxx.xxx.xxx.xxx
time="2023-12-27T14:05:33+01:00" level=debug msg="Authorization Request with id '49da15e9-b23e-45a7-8ba9-4d94f14770e8' on client with id 'centreon' is being processed" method=GET path=/api/oidc/authorization remote_ip=xxxx.xxx.xxx.xxx
time="2023-12-27T14:05:33+01:00" level=debug msg="Authorization Request with id '49da15e9-b23e-45a7-8ba9-4d94f14770e8' on client with id 'centreon' was successfully processed, proceeding to build Authorization Response" method=GET path=/api/oidc/authorization remote_ip=xxxx.xxx.xxx.xxx
time="2023-12-27T14:05:33+01:00" level=trace msg="Authorization Request with id '49da15e9-b23e-45a7-8ba9-4d94f14770e8' on client with id 'centreon' creating session for Authorization Response for subject '5b851fb2-2b94-4c0e-a6bf-6d1d00b343a3' with username 'fabien' with claims: &{JTI: Issuer:https://authelia.xxxx-xxx.fr Subject:5b851fb2-2b94-4c0e-a6bf-6d1d00b343a3 Audience:[centreon] Nonce: ExpiresAt:0001-01-01 00:00:00 +0000 UTC IssuedAt:2023-12-27 14:05:33.487480757 +0100 CET m=+759.555731094 RequestedAt:2023-12-27 13:05:32.337898962 +0000 UTC AuthTime:2023-12-27 13:59:24 +0100 CET AccessTokenHash: AuthenticationContextClassReference: AuthenticationMethodsReferences:[pwd otp mfa] CodeHash: Extra:map[azp:centreon client_id:centreon name:Fabien FORTON preferred_username:fabien]}" method=GET path=/api/oidc/authorization remote_ip=xxxx.xxx.xxx.xxx
time="2023-12-27T14:05:33+01:00" level=trace msg="Replied (status=303)" method=GET path=/api/oidc/authorization remote_ip=xxxx.xxx.xxx.xxx
time="2023-12-27T14:05:33+01:00" level=trace msg="Request hit" method=POST path=/api/oidc/token remote_ip=xxxx.xxx.xxx.xxx
time="2023-12-27T14:05:33+01:00" level=error msg="Access Request failed with error: The provided authorization grant (e.g., authorization code, resource owner credentials) or refresh token is invalid, expired, revoked, does not match the redirection URI used in the authorization request, or was issued to another client. The 'redirect_uri' from this request does not match the one from the authorize request." method=POST path=/api/oidc/token remote_ip=xxxx.xxx.xxx.xxx stack="github.com/authelia/authelia/v4/internal/handlers/handler_oidc_token.go:27                   OpenIDConnectTokenPOST\ngithub.com/authelia/authelia/v4/internal/middlewares/http_to_authelia_handler_adaptor.go:113 NewHTTPToAutheliaHandlerAdaptor.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/bridge.go:54                            (*BridgeBuilder).Build.func1.1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:35                           SecurityHeadersNoStore.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:25                           SecurityHeadersCSPNone.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:16                           SecurityHeaders.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/cors.go:216                             (*CORSPolicy).Middleware.func1\ngithub.com/fasthttp/router@v1.4.14/router.go:414                                             (*Router).Handler\ngithub.com/valyala/fasthttp@v1.43.0/http.go:154                                              (*Response).StatusCode\ngithub.com/authelia/authelia/v4/internal/middlewares/strip_path.go:22                        StripPath.func1.1\ngithub.com/valyala/fasthttp@v1.43.0/server.go:2338                                           (*Server).serveConn\ngithub.com/valyala/fasthttp@v1.43.0/workerpool.go:224                                        (*workerPool).workerFunc\ngithub.com/valyala/fasthttp@v1.43.0/workerpool.go:196                                        (*workerPool).getCh.func1\nruntime/asm_amd64.s:1594                                                                     goexit"
time="2023-12-27T14:05:33+01:00" level=trace msg="Replied (status=400)" method=POST path=/api/oidc/token remote_ip=xxxx.xxx.xxx.xxx

**Reply by james-d-elliott** [reply]
Looks fine though I don't know anything about it. What have they said about the error? It sounds like maybe they can't make requests to Authelia's token URL but I don't know their errors, so without Authelia logs we're just guessing.

**Reply by james-d-elliott** [reply]
Looks like they are not providing a `redirect_uri` in the request to the token endpoint or it is different to the one in the authorize request which is a bug on their end if so. I'm happy to create a method for you to confirm this but it's probably an issue on their end.

See https://datatracker.ietf.org/doc/html/rfc6749#section-4.1.3

```
redirect_uri
         REQUIRED, if the "redirect_uri" parameter was included in the
         authorization request as described in [Section 4.1.1](https://datatracker.ietf.org/doc/html/rfc6749#section-4.1.1), and their
         values MUST be identical.
```

**Reply by g4fa** [reply]
You think it is an issue of centreon ?

**Reply by james-d-elliott** [reply, marked as answer]
Seems very likely a bug on their end. Note the rtrim function being applied to one of the Redirect URI's, also looks like they are generated potentially randomly? That being said I'm not a PHP developer.  It could be looking up a value from config or the request, either way they must be identical to be compliant with the spec.

This is the value sent to the authorization endpoint:

https://github.com/centreon/centreon/blob/57465898ae16c2b3c68119118bee36e2aaecec58/centreon/src/Core/Security/ProviderConfiguration/Infrastructure/Api/FindProviderConfigurations/ProviderPresenter/OpenIdProviderPresenter.php#L110-L115

This is the value sent to the token endpoint:

https://github.com/centreon/centreon/blob/57465898ae16c2b3c68119118bee36e2aaecec58/centreon/src/Core/Security/Authentication/Domain/Provider/OpenIdProvider.php#L456-L460

**Reply by g4fa** [reply]
Can you try in your homelab to configure Authelia with centreon ? 
Thanks for your help
